File: app/lib/services/shelf_life.dart
class ShelfLifeService {
  static const Map<String, int> baseShelfLife = {
    'banana': 5,
    'bittermelon': 4,
    'cucumber': 7,
    'eggplant': 5,
    'orange': 14,
    'papaya': 5,
    'pineapple': 3,
    'tomato': 7
  };

  static const Map<String, double> conditionModifiers = {
    'Fresh': 1.0,
    'Semi Fresh': 0.6,
    'Rotten': 0.0
  };

  static int estimateShelfLife(String fruit, String freshness, double tempC, double humidityPercent) {
    if (freshness == 'Unknown') {
      return 0; // Not determinable
    }

    String normalizedFruit = fruit.toLowerCase();
    double baseDays = (baseShelfLife[normalizedFruit] ?? 5).toDouble();
    double conditionMod = conditionModifiers[freshness] ?? 0.0;

    if (conditionMod == 0.0) return 0; // Rotten

    // Temperature Logic (Ideal ~15C)
    double tempMod = 1.0;
    if (tempC > 25) {
      tempMod = 0.5;
    } else if (tempC > 20) {
      tempMod = 0.8;
    } else if (tempC < 10) {
      tempMod = 1.2;
    }

    // Humidity Logic (Ideal ~85%)
    double humidityMod = 1.0;
    if (humidityPercent > 90) {
      humidityMod = 0.8; // Too humid, rot faster
    } else if (humidityPercent < 60) {
      humidityMod = 0.7; // Too dry, shrivel faster
    }

    double rawEstimate = baseDays * conditionMod * tempMod * humidityMod;
    int finalEstimate = rawEstimate.round();

    return finalEstimate < 1 ? 1 : finalEstimate;
  }
}

File: frontend/src/lib/shelfLife.ts
export function estimateShelfLife(fruit: string, freshness: string, temperature: number, humidity: number): string {
  const fallbackShelfLife: Record<string, Record<string, [number, number]>> = {
    'banana': {'fresh': [5, 7], 'semi_fresh': [2, 3], 'rotten': [0, 0]},
    'orange': {'fresh': [10, 14], 'semi_fresh': [3, 5], 'rotten': [0, 0]},
    'papaya': {'fresh': [4, 6], 'semi_fresh': [1, 2], 'rotten': [0, 0]},
    'pineapple': {'fresh': [3, 5], 'semi_fresh': [1, 2], 'rotten': [0, 0]},
    'tomato': {'fresh': [7, 10], 'semi_fresh': [2, 4], 'rotten': [0, 0]},
    'cucumber': {'fresh': [7, 10], 'semi_fresh': [2, 3], 'rotten': [0, 0]},
    'eggplant': {'fresh': [5, 7], 'semi_fresh': [1, 3], 'rotten': [0, 0]},
    'bittermelon': {'fresh': [4, 6], 'semi_fresh': [1, 2], 'rotten': [0, 0]}
  };
  
  const fruitKey = fruit.toLowerCase().replace(/ /g, '_');
  const freshKey = freshness.toLowerCase().replace(/ /g, '_');
  
  if (!fallbackShelfLife[fruitKey]) {
      return "Unknown / Needs Additional Model Support";
  }
      
  const baseLife = fallbackShelfLife[fruitKey][freshKey] || [1, 3];
  
  // Environmental factors
  let multiplier = 1.0;
  if (temperature > 25) {
      multiplier *= Math.pow(0.95, temperature - 25);
  }
  
  // Humidity logic (simple heuristic)
  if (humidity < 40) {
      multiplier *= 0.9; // too dry, might shrivel
  } else if (humidity > 85) {
      multiplier *= 0.9; // too humid, might mold fast
  }
      
  const minDays = Math.max(0, Math.round(baseLife[0] * multiplier));
  const maxDays = Math.max(minDays, Math.round(baseLife[1] * multiplier));
  
  if (minDays === 0 && maxDays === 0) {
      return "0 Days (Discard)";
  }
  return `${minDays}–${maxDays} Days`;
}
